File: pattern_matching_v2/generate_test.py
import random
import argparse

def generate_random_row(length, charset):
    if length == 0: return ""
    return "".join(random.choice(charset) for _ in range(length))

def generate_pattern(m1, m2, charset):
    if m1 == 0: return []
    return [generate_random_row(m2, charset) for _ in range(m1)]

def generate_text_background(n1, n2, charset):
    if n1 == 0: return []
    return [generate_random_row(n2, charset) for _ in range(n1)]

def get_lps_for_rows(pattern_rows):
    m1 = len(pattern_rows)
    if m1 == 0:
        return []
    lps = [0] * m1
    length = 0
    i = 1
    while i < m1:
        if pattern_rows[i] == pattern_rows[length]:
            length += 1
            lps[i] = length
            i += 1
        else:
            if length != 0:
                length = lps[length - 1]
            else:
                lps[i] = 0
                i += 1
    return lps

def plant_pattern_in_text(text_rows_list_of_list_of_chars, pattern, r_plant, c_plant):
    m1_p = len(pattern)
    if m1_p == 0:
        return
    m2_p = len(pattern[0])

    for i in range(m1_p):
        for j in range(m2_p):
            if r_plant + i < len(text_rows_list_of_list_of_chars) and \
                    c_plant + j < len(text_rows_list_of_list_of_chars[r_plant + i]):
                text_rows_list_of_list_of_chars[r_plant + i][c_plant + j] = pattern[i][j]

def check_rect_overlap(r1, c1, height1, width1, r2, c2, height2, width2):
    if (c1 + width1 <= c2 or
            c2 + width2 <= c1 or
            r1 + height1 <= r2 or
            r2 + height2 <= r1):
        return False
    return True

def generate_test_case(m1, m2, n1, n2, charset, num_target_occurrences, overlap_strategy):
    if m1 == 0:
        P = []
        T_bg = generate_text_background(n1, n2, charset)
        return P, T_bg, []

    P = generate_pattern(m1, m2, charset)
    T_bg_rows = generate_text_background(n1, n2, charset)

    T_list_of_list_chars = [list(row) for row in T_bg_rows]

    planted_occurrences_coords = []

    lps_P_rows = get_lps_for_rows(P)
    kmp_overlap_vertical_shift = 0
    if m1 > 0 and lps_P_rows[m1 - 1] > 0:
        kmp_overlap_vertical_shift = m1 - lps_P_rows[m1 - 1]

    possible_initial_coords = []
    if m1 > 0 :
        for r in range(n1 - m1 + 1):
            for c in range(n2 - m2 + 1):
                possible_initial_coords.append((r, c))

    random.shuffle(possible_initial_coords)

    if not possible_initial_coords and num_target_occurrences > 0:
        T_final_rows = ["".join(row_chars) for row_chars in T_list_of_list_chars]
        return P, T_final_rows, []


    if overlap_strategy == "kmp_driven" and kmp_overlap_vertical_shift > 0 and \
            num_target_occurrences >= 2 and len(possible_initial_coords) > 0:

        r1, c1 = possible_initial_coords.pop(0)
        plant_pattern_in_text(T_list_of_list_chars, P, r1, c1)
        planted_occurrences_coords.append((r1, c1))

        r2, c2 = r1 + kmp_overlap_vertical_shift, c1
        if (r2 + m1 <= n1 and c2 + m2 <= n2):
            if (r2,c2) != (r1,c1):
                plant_pattern_in_text(T_list_of_list_chars, P, r2, c2)
                if (r2,c2) not in planted_occurrences_coords:
                    planted_occurrences_coords.append((r2, c2))
                if (r2,c2) in possible_initial_coords:
                    possible_initial_coords.remove((r2,c2))

    num_already_planted = len(planted_occurrences_coords)
    num_to_plant_additionally = num_target_occurrences - num_already_planted

    additional_planted_this_loop = 0

    coords_candidates_for_loop = list(possible_initial_coords)

    for r_cand, c_cand in coords_candidates_for_loop:
        if additional_planted_this_loop >= num_to_plant_additionally:
            break
        if len(planted_occurrences_coords) >= num_target_occurrences:
            break

        can_plant = True
        if overlap_strategy == "none":
            for r_prev, c_prev in planted_occurrences_coords:
                if check_rect_overlap(r_cand, c_cand, m1, m2, r_prev, c_prev, m1, m2):
                    can_plant = False
                    break

        if can_plant:
            if (r_cand, c_cand) not in planted_occurrences_coords:
                plant_pattern_in_text(T_list_of_list_chars, P, r_cand, c_cand)
                planted_occurrences_coords.append((r_cand, c_cand))
                additional_planted_this_loop += 1

    planted_occurrences_coords.sort()

    T_final_rows = ["".join(row_chars) for row_chars in T_list_of_list_chars]

    output_occurrences_1_indexed = sorted([(r + 1, c + 1) for r, c in planted_occurrences_coords])
    return P, T_final_rows, output_occurrences_1_indexed


def main():
    parser = argparse.ArgumentParser(description="Generate test cases for 2D KMP-like algorithm.")
    parser.add_argument("--m1", type=int, default=2, help="Number of rows in pattern P")
    parser.add_argument("--m2", type=int, default=2, help="Number of columns in pattern P")
    parser.add_argument("--n1", type=int, default=8, help="Number of rows in text T")
    parser.add_argument("--n2", type=int, default=8, help="Number of columns in text T")
    parser.add_argument("--charset", type=str, default="AB", help="Characters to use for generation")
    parser.add_argument("--num_occurrences", type=int, default=2, help="Target number of P occurrences to plant in T")
    parser.add_argument("--overlap_strategy", type=str, choices=["random", "kmp_driven", "none"], default="random",
                        help="Strategy for planting patterns")
    parser.add_argument("--seed", type=int, default=None, help="Random seed for reproducibility")
    parser.add_argument("--output_file", type=str, default=None, help="File to write the test case to")

    args = parser.parse_args()

    if args.seed is not None:
        random.seed(args.seed)

    if args.m1 < 0 or args.m2 < 0 or args.n1 < 0 or args.n2 < 0 :
        print("Error: Dimensions (m1, m2, n1, n2) must be non-negative.")
        return
    if args.m1 == 0 and args.num_occurrences > 0 :
        print("Warning: m1 is 0 (no pattern rows). Setting num_occurrences to 0.")
        args.num_occurrences = 0
    if args.m1 > 0 and args.m2 == 0 and args.num_occurrences > 0:
        print("Warning: Pattern has m2=0 (0 columns).")

    P, T, expected_occurrences = generate_test_case(
        args.m1, args.m2, args.n1, args.n2,
        args.charset, args.num_occurrences, args.overlap_strategy
    )

    output_lines = []
    output_lines.append(f"{args.m1} {args.m2} {args.n1} {args.n2}")
    for p_row in P:
        output_lines.append(p_row)
    for t_row in T:
        output_lines.append(t_row)

    solution_lines = []
    solution_lines.append(f"\n# --- Expected Solution (Intentionally Planted Occurrences) ---")
    solution_lines.append(f"# Count: {len(expected_occurrences)}")
    for r_out, c_out in expected_occurrences:
        solution_lines.append(f"# {r_out} {c_out}")

    if args.output_file:
        with open(args.output_file, 'w') as f:
            for line in output_lines:
                f.write(line + '\n')
            for line in solution_lines:
                f.write(line + '\n')
        print(f"Test case input and expected solution comments written to {args.output_file}")
    else:
        for line in output_lines:
            print(line)
        for line in solution_lines:
            print(line)

if __name__ == "__main__":
    main()

File: pattern_matching_v2/generate_test.cpp
#include <iostream>
#include <vector>
#include <string>
#include <random>
#include <algorithm>
#include <fstream>
#include <utility>
#include <optional>
#include <set>
#include <map>
#include <stdexcept>

using Matrix = std::vector<std::string>;
using Coords = std::pair<int, int>;

std::string generate_random_row(int length, const std::string& charset, std::mt19937& gen) {
    if (length == 0) return "";
    std::uniform_int_distribution<> distrib(0, charset.length() - 1);
    std::string row;
    row.reserve(length);
    for (int i = 0; i < length; ++i) {
        row += charset[distrib(gen)];
    }
    return row;
}

Matrix generate_matrix(int rows, int cols, const std::string& charset, std::mt19937& gen) {
    if (rows == 0) return {};
    Matrix mat;
    mat.reserve(rows);
    for (int i = 0; i < rows; ++i) {
        mat.push_back(generate_random_row(cols, charset, gen));
    }
    return mat;
}

std::vector<int> get_lps_for_rows(const Matrix& pattern_rows) {
    int m1 = pattern_rows.size();
    if (m1 == 0) return {};
    
    std::vector<int> lps(m1, 0);
    int length = 0;
    int i = 1;
    while (i < m1) {
        if (pattern_rows[i] == pattern_rows[length]) {
            length++;
            lps[i] = length;
            i++;
        } else {
            if (length != 0) {
                length = lps[length - 1];
            } else {
                lps[i] = 0;
                i++;
            }
        }
    }
    return lps;
}

void plant_pattern_in_text(Matrix& text, const Matrix& pattern, int r_plant, int c_plant) {
    if (pattern.empty()) return;
    
    int m1_p = pattern.size();
    int m2_p = pattern[0].size();
    
    for (int i = 0; i < m1_p; ++i) {
        int target_row = r_plant + i;
        if (target_row >= text.size()) continue;
        for (int j = 0; j < m2_p; ++j) {
            int target_col = c_plant + j;
            if (target_col < text[target_row].size()) {
                text[target_row][target_col] = pattern[i][j];
            }
        }
    }
}

bool check_rect_overlap(int r1, int c1, int h1, int w1, int r2, int c2, int h2, int w2) {
    if (c1 + w1 <= c2 || c2 + w2 <= c1 || r1 + h1 <= r2 || r2 + h2 <= r1) {
        return false;
    }
    return true;
}

struct TestCaseResult {
    Matrix P;
    Matrix T;
    std::vector<Coords> occurrences;
};

TestCaseResult generate_test_case(int m1, int m2, int n1, int n2, const std::string& charset,
                                    int num_target_occurrences, const std::string& overlap_strategy, std::mt19937& gen) {
    if (m1 == 0) {
        Matrix T_bg = generate_matrix(n1, n2, charset, gen);
        return {{}, T_bg, {}};
    }

    Matrix P = generate_matrix(m1, m2, charset, gen);
    Matrix T = generate_matrix(n1, n2, charset, gen);
    std::vector<Coords> planted_occurrences_coords;
    std::set<Coords> used_coords;

    std::vector<Coords> possible_initial_coords;
    if (m1 > 0 && n1 >= m1 && n2 >= m2) {
        for (int r = 0; r <= n1 - m1; ++r) {
            for (int c = 0; c <= n2 - m2; ++c) {
                possible_initial_coords.emplace_back(r, c);
            }
        }
    }
    std::shuffle(possible_initial_coords.begin(), possible_initial_coords.end(), gen);

    if (possible_initial_coords.empty() && num_target_occurrences > 0) {
        return {P, T, {}};
    }

    if (overlap_strategy == "kmp_driven" && num_target_occurrences >= 2 && !possible_initial_coords.empty()) {
        auto lps_P_rows = get_lps_for_rows(P);
        int kmp_overlap_vertical_shift = m1 > 0 && lps_P_rows.back() > 0 ? m1 - lps_P_rows.back() : 0;
        
        if (kmp_overlap_vertical_shift > 0) {
            Coords r1c1 = possible_initial_coords.front();
            plant_pattern_in_text(T, P, r1c1.first, r1c1.second);
            planted_occurrences_coords.push_back(r1c1);
            used_coords.insert(r1c1);

            Coords r2c2 = {r1c1.first + kmp_overlap_vertical_shift, r1c1.second};
            if (r2c2.first + m1 <= n1 && r2c2.second + m2 <= n2 && used_coords.find(r2c2) == used_coords.end()) {
                plant_pattern_in_text(T, P, r2c2.first, r2c2.second);
                planted_occurrences_coords.push_back(r2c2);
                used_coords.insert(r2c2);
            }
        }
    }

    for (const auto& coord_cand : possible_initial_coords) {
        if (planted_occurrences_coords.size() >= num_target_occurrences) break;
        if (used_coords.count(coord_cand)) continue;

        bool can_plant = true;
        if (overlap_strategy == "none") {
            for (const auto& prev_coord : planted_occurrences_coords) {
                if (check_rect_overlap(coord_cand.first, coord_cand.second, m1, m2, 
                                        prev_coord.first, prev_coord.second, m1, m2)) {
                    can_plant = false;
                    break;
                }
            }
        }
        
        if (can_plant) {
            plant_pattern_in_text(T, P, coord_cand.first, coord_cand.second);
            planted_occurrences_coords.push_back(coord_cand);
            used_coords.insert(coord_cand);
        }
    }

    std::sort(planted_occurrences_coords.begin(), planted_occurrences_coords.end());
    std::vector<Coords> output_occurrences_1_indexed;
    for (const auto& coord : planted_occurrences_coords) {
        output_occurrences_1_indexed.emplace_back(coord.first + 1, coord.second + 1);
    }
    
    return {P, T, output_occurrences_1_indexed};
}

int main(int argc, char* argv[]) {
    std::map<std::string, std::string> args;
    args["--m1"] = "2";
    args["--m2"] = "2";
    args["--n1"] = "8";
    args["--n2"] = "8";
    args["--charset"] = "AB";
    args["--num_occurrences"] = "2";
    args["--overlap_strategy"] = "random";
    args["--seed"] = "";
    args["--output_file"] = "";

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (args.count(arg) && i + 1 < argc) {
            args[arg] = argv[++i];
        }
    }
    
    int m1, m2, n1, n2, num_occurrences;
    try {
        m1 = std::stoi(args["--m1"]);
        m2 = std::stoi(args["--m2"]);
        n1 = std::stoi(args["--n1"]);
        n2 = std::stoi(args["--n2"]);
        num_occurrences = std::stoi(args["--num_occurrences"]);
    } catch (const std::invalid_argument& e) {
        std::cerr << "Error: Invalid number provided for dimensions or occurrences." << std::endl;
        return 1;
    }
    
    std::string charset = args["--charset"];
    std::string overlap_strategy = args["--overlap_strategy"];
    std::optional<unsigned int> seed;
    if (!args["--seed"].empty()) {
        try {
            seed = std::stoul(args["--seed"]);
        } catch (const std::invalid_argument& e) {
            std::cerr << "Error: Invalid number provided for seed." << std::endl;
            return 1;
        }
    }
    std::string output_file = args["--output_file"];

    if (m1 < 0 || m2 < 0 || n1 < 0 || n2 < 0) {
        std::cerr << "Error: Dimensions (m1, m2, n1, n2) must be non-negative." << std::endl;
        return 1;
    }
    if (m1 == 0 && num_occurrences > 0) {
        std::cerr << "Warning: m1 is 0 (no pattern rows). Setting num_occurrences to 0." << std::endl;
        num_occurrences = 0;
    }
    if (m1 > 0 && m2 == 0 && num_occurrences > 0) {
        std::cerr << "Warning: Pattern has m2=0 (0 columns)." << std::endl;
    }

    std::mt19937 gen;
    if (seed.has_value()) {
        gen.seed(seed.value());
    } else {
        std::random_device rd;
        gen.seed(rd());
    }

    TestCaseResult result = generate_test_case(m1, m2, n1, n2, charset, num_occurrences, overlap_strategy, gen);
    
    std::vector<std::string> output_lines;
    output_lines.push_back(std::to_string(m1) + " " + std::to_string(m2) + " " + std::to_string(n1) + " " + std::to_string(n2));
    for (const auto& p_row : result.P) {
        output_lines.push_back(p_row);
    }
    for (const auto& t_row : result.T) {
        output_lines.push_back(t_row);
    }

    std::vector<std::string> solution_lines;
    solution_lines.push_back("\n# --- Expected Solution (Intentionally Planted Occurrences) ---");
    solution_lines.push_back("# Count: " + std::to_string(result.occurrences.size()));
    for (const auto& coord : result.occurrences) {
        solution_lines.push_back("# " + std::to_string(coord.first) + " " + std::to_string(coord.second));
    }

    if (!output_file.empty()) {
        std::ofstream outfile(output_file);
        if (!outfile) {
            std::cerr << "Error: Could not open file for writing: " << output_file << std::endl;
            return 1;
        }
        for (const auto& line : output_lines) {
            outfile << line << '\n';
        }
        for (const auto& line : solution_lines) {
            outfile << line << '\n';
        }
        std::cout << "Test case input and expected solution comments written to " << output_file << std::endl;
    } else {
        for (const auto& line : output_lines) {
            std::cout << line << '\n';
        }
        for (const auto& line : solution_lines) {
            std::cout << line << '\n';
        }
    }

    return 0;
}
